### ISSUE TITLE
NPC Rabbit

### ISSUE THREAD

**Issue by R0TX** [RotX18/ARProject#1]
## Objective: Add a simple moving rabbit with animals onto the scene

Rabbit should use invisible empty game objects as waypoints to move around, it should face the direction it is moving in

Will try to make so that the rabbit can approach the player if it is facing the player after basic movement and behavior of the rabbit is done.

Ultimately, plan to add features such as "feeding" and "petting" to award points to conform to lyf's point system

### Required:
- [x] Rabbit Asset
- [x] Movement Script
- [x] Waypoints (Empty game objects)
- [x] Petting Feature

**Comment by R0TX** [comment]
# Started work on implementing rabbit
Added rabbit model and Waypoints

## Rabbit model[^animations] and waypoints:
![image](https://user-images.githubusercontent.com/84174550/167349792-00eaa2f2-7058-4fd7-a5cb-6f5034c8481b.png)

Waypoints are indicated by the blue spheres

## Rabbit movement script:
Added the rabbit movement script to the rabbit called RabbitBehavior.cs, plan to add more behavior into this script such as the rabbit noticing and turning toward the player [^behavior] 
Script is as follows:

### RabbitBehavior.cs
```cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class RabbitBehavior : MonoBehaviour
{
    //PUBLIC VARS
    public GameObject[] waypoints;

    //PRIVATE VARS
    private GameObject targetPos;

    private void Start() {
        //at Start, select a random waypoint to move toward
        targetPos = waypoints[Random.Range(0, waypoints.Length)];
    }

    private void Update() {
        //changing the position to the target position using Vector3.MoveTowards();
        transform.position = Vector3.MoveTowards(transform.position, targetPos.transform.position, 0.005f);

        if(transform.position == targetPos.transform.position) {
            //once rabbit has reached the target waypoint, change the target position so that it can move toward the new target
            SetNewTarget();
        }
    }

    private void SetNewTarget() {
        GameObject newTarget = waypoints[Random.Range(0, waypoints.Length)];

        if(targetPos.transform.position == newTarget.transform.position) {
            //rerolling if the new position is the same as the current position
            targetPos = waypoints[Random.Range(0, waypoints.Length)];
        }
        else {
            targetPos = newTarget;
        }
    }
}
```
In Start(), the script chooses a random waypoint to send the rabbit to. 

Next in Update(), the rabbit moves toward the target position. When the rabbit reaches the target position, it selects a new position through the use of SetNewTarget().

SetNewTarget() selects a waypoint at random and sets the targetPos global GameObject to the position of the selected waypoint, reselecting if the waypoint is the same as the current position. 

https://user-images.githubusercontent.com/84174550/167351791-faf565df-2126-4534-ae40-97125d432208.mp4

May change the position of the waypoints and add more in the future to give a more natural vibe

[^animations]: Animations and the rabbit turning are not done as of this comment, will add them later
[^behavior]: Additional behavior such as  the aforementioned rabbit noticing the player will either be added later down the line or scrapped if deemed overly complicated

**Comment by R0TX** [comment]
# Added rabbit behaviour and animations + Miscellaneous changes
Rabbits now face the direction that they are moving and have an idle and running animation

## New behavior: idle, movement and facing
Added and changed RabbitMovement.cs and added animation to rabbit models. Changes are as follows:

### RabbitMovement.cs (changed code)
```cs
    //PUBLIC VARS
    public Animator anim;

    //PRIVATE VARS
    //floats
    private float _moveSpeed = 0.003f;
    private float _idleChance = 1;

    //bools
    private bool _executeCoroutines = true;
    private bool _move = true;

    private void Update() {
        //changing the position to the target position using Vector3.MoveTowards();
        if(_move){
            transform.position = Vector3.MoveTowards(transform.position, _targetPos.transform.position, _moveSpeed);
            transform.LookAt(_targetPos.transform);
        }

        if(transform.position == _targetPos.transform.position) {
            //once rabbit has reached the target waypoint check whether it should stay in place for a few seconds
            if(Random.Range(1, _idleChance) == 1 && _executeCoroutines == true){
                //there is a 1 in _idleChance that the rabbit will stay idle for awhile
                _move = false;
                _executeCoroutines = false;
                StartCoroutine(goIdle());
            }
            else{
                SetNewTarget();
            }
        }
    }

    private IEnumerator goIdle(){
        //coroutine to set the animator to idle state for a random amt of seconds before reverting
        anim.SetBool("IdleState", true);
        yield return new WaitForSeconds(Random.Range(3, 7));
        anim.SetBool("IdleState", false);

        //allowing this coroutine to be executed again
        _executeCoroutines = true;

        //setting new target and allowing the rabbit to move
        SetNewTarget();
        _move = true;
    }
```
Edits:
- Added:
  - Animator to control rabit animation
  - Floats responsible for movement and the chance of idling[^idleChance]
  - Booleans to restrict movement and execute idle state
  - Coroutine to allow for idling
- Changed:
  - Update function:
     - Added transform.LookAt to movement so that rabbit will face the direction it is moving
     - Added an if statement to check whether to idle based on the value of a Random.Range

### Rabbit Animator
![image](https://user-images.githubusercontent.com/84174550/169388658-c9e9e9c8-9a1f-4692-a283-e255531e3607.png)
Notes:
- IdleState is the condition for transitioning to and from the two states shown
- Values within the animations and transitions may change

### Updated rabbits
https://user-images.githubusercontent.com/84174550/169390642-29e1fe53-e1c0-4c0e-af54-4998d217f28e.mp4

Rabbits move naturally and randomly, successfully achieved the goal of attaining natural rabbit movement
## Miscellaneous changes
Aside from the changes to code, I have also made some miscellaneous changes to the project

### Prefabs and waypoints [^waypoints]
Added and renamed 6 more waypoints
![image](https://user-images.githubusercontent.com/84174550/169391797-8136b3b0-a13c-4b6a-87a9-3f296c9748db.png)

### Reorganised hierachy to improve tidiness and allow for better quality of life
![image](https://user-images.githubusercontent.com/84174550/169392453-350c74c4-2540-4cae-b883-1f5fda19feb0.png)
Objects are now sorted in alphabetical order to allow for easy searching and identification

### Standardised heights for objects in the scene
![image](https://user-images.githubusercontent.com/84174550/169394275-0c35b6ea-591c-41ca-98bb-3e994ccd195e.png)
Most if not all objects appear to be touching the floor now

> Will continue to ideate and think of improvements and enhancements as I finish up NPC-Rabbits

[^idleChance]: For testing purposes, _idleChance is set to 1, but idling chance may be removed entirely if deemed unnecessary/redundant
[^waypoints]: Each rabbit uses different waypoints (odd numbered vs even numbered) to reduce the chances of them travelling through each other, waypoints will be changed to empty game objects before building

**Comment by R0TX** [comment]
# Added foundation for the petting function
In theory, I should:
1. Raycast forward relative to the camera's position
2. Compare the tag to check whether there is a Rabbit tag to the hit game object
3. if true, enable the pet button
4. Play the heart particle system and Debug.Log("You got 1 coin!");

Changes to the code are as follows:

## Added CameraInteraction.cs to handle the raycasting
Still buggy but as mentioned, this is skeleton code

### CameraInteraction.cs:
```cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.UI;

public class CameraInteraction : MonoBehaviour
{
    //PUBLIC VARS
    public Button btnPet;

    //PRIVATE VARS
    private RaycastHit _hit;
    private GameObject _rabbitParticles;

    private void FixedUpdate() {
        Debug.DrawRay(transform.position, Vector3.forward, Color.red);
        if(Physics.Raycast(transform.position, Vector3.forward, out _hit)){
            //raycasting from the camera to 3f units infront
            if(_hit.transform.CompareTag("Rabbit")){
                //if the raycast obj has rabbit as its tag
                btnPet.gameObject.SetActive(true);

                //getting the game object responsible for the particles
                _rabbitParticles = _hit.transform.GetChild(3).gameObject;
            }
        }
        else{
            btnPet.gameObject.SetActive(false);
        }
    }

    public void BtnPetOnClick(){
        //playing the hearts particle and debug logging the coin
        _rabbitParticles.GetComponent<ParticleSystem>().Play();
        Debug.Log("You have earned 1 coin!");
    }
}
```
Issues: 
- Raycasting does not seem to work, code does not set the button to true, thus the rest cannot be tested
- Debug statements not showing in console, will try to solve before proceeding

Will attempt to fix / seek assistance on this matter

## Miscellaneous: Updated Vuphoria to 10.7.2 and Imported RainMaker asset
### Vuphoria package updated

### RainMaker asset imported in preparation to add rain in the app (#2)
Details on implementation can be found in the respective issue

**Comment by R0TX** [comment]
# Fixed issue with Raycasting
Fixed issues with raycasting, Petting feature now works on rabbit

## Edited code to work with raycasting and particle system
Fixed bug where raycasting was not working
### CameraInteraction.cs
```cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.UI;

public class CameraInteraction : MonoBehaviour
{
    //PUBLIC VARS
    public Button btnPet;

    //PRIVATE VARS
    private RaycastHit _hit;
    private ParticleSystem _rabbitParticles;

    private void FixedUpdate() {
        if(Physics.Raycast(transform.position, transform.forward, out _hit, Mathf.Infinity)){
            //raycasting from the camera to 3f units infront
            if(_hit.transform.CompareTag("Rabbit")){
                //if the raycast obj has rabbit as its tag
                btnPet.gameObject.SetActive(true);

                //getting the game object responsible for the particles
                _rabbitParticles = _hit.transform.GetComponent<ParticleSystem>();
            }
        }
        else{
            btnPet.gameObject.SetActive(false);
        }
    }

    public void OnClick_BtnPet(){
        //playing the hearts particle and debug logging the coin
        _rabbitParticles.Play();
        Debug.Log("You have earned 1 coin!");
    }
}
```
Edits:
- Changed code to use transform.forward and Mathf.Infinity in Raycast direction and maxDistance respectively to allow camera to target rabbit
- Added particle system component to rabbit so that it is part of the rabbit as GetComponentInChildren\<ParticleSystem\>() was returning null

**Comment by R0TX** [comment]
# AR Application Demonstration

https://user-images.githubusercontent.com/84174550/173448889-fa0af473-6451-4e8a-9a23-645b1a8e4b19.mp4

> Feature has been implemented, issue can now be closed, will merge after (pull request: #5 )
